### ISSUE TITLE
How to remove the randomness to get reproducible runs

### ISSUE THREAD

**Issue by a-mos** [BachiLi/diffvg#63]
Hello! I am trying to get fully-reproducible pipeline, however with all seeds fixed i got different grad values on first iteration in single_circle.py app even with same loss value:

```
Fixing all seeds:

import random
import os

seed = 0
torch.set_printoptions(precision=16)
os.environ['PYTHONHASHSEED'] = str(seed)
torch.manual_seed(seed)
random.seed(seed)
np.random.seed(seed)
torch.backends.cudnn.benchmark = False
torch.use_deterministic_algorithms(True)
os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.enabled = False
torch.cuda.manual_seed(seed)
torch.cuda.manual_seed_all(seed)
torch.set_deterministic_debug_mode(2)
```

Run1 python single_circle.py
```
iteration: 0
loss: 6360.21044921875
radius.grad: tensor(-6679.3833007812500000)
center.grad: tensor([-16965.3789062500000000,   8691.6396484375000000])
```

Run2 python single_circle.py
```
iteration: 0
loss: 6360.21044921875
radius.grad: tensor(-6679.9208984375000000)
center.grad: tensor([-16965.9726562500000000,   8691.4550781250000000])
```

**Comment by tanguymagne** [comment]
Hello,
I am facing the same issue. Did you manage to solve it?
Thank you very much

**Comment by L483** [comment]
Any luck so far? I have essentially the same problem, but inside `painterly_rendering.py`
I added most of the same lines in my efforts to make it deterministic.
The initial PNGs and SVGs before the first backwards pass can be reproduced, but the following iterations diverge more and more between runs.

I assume that there is non-deterministic code inside `diffvg.cpp` or other C++ parts, but I am not well-versed in C++. At the very least, the non-deterministic behavior should not stem from the python code itself.

**Comment by a-mos** [comment]
> Any luck so far? I have essentially the same problem, but inside `painterly_rendering.py` I added most of the same lines in my efforts to make it deterministic. The initial PNGs and SVGs before the first backwards pass can be reproduced, but the following iterations diverge more and more between runs.
> 
> I assume that there is non-deterministic code inside `diffvg.cpp` or other C++ parts, but I am not well-versed in C++. At the very least, the non-deterministic behavior should not stem from the python code itself.

Hello, unfortunately, no, the randomness was not on the side of Pytorch but directly in the diffvg code.

**Comment by L483** [comment]
Yeah, the randomness is caused by the parallel float additions within the same memory locations. The order in which threads add their respective values is not deterministic, and the associative property does not apply here, which causes the randomness. Only `render_edge_kernel` and `render_kernel` suffer from this, as far as I have tested.

Making the C++ single-threaded removes randomness, but also, when tested in that way, a single iteration takes ~1150 times longer (I calculated it solely with the logged forward and backward passes, but also took just a single sample for this calculation).
Providing each thread with its own memory to write to and accumulating their updates deterministically (inside the `parallel_for`) would need ~160GB of memory (because it spawns ~200.000 threads) for a `224x224`px image if I did understand the code correctly.

Depending on the image size, a viable approach could be to not spawn a new CUDA thread per work unit, but only for every few work units, so there are few enough threads to provide each one with its own memory space to write to. Each thread would act deterministically, and accumulating their updates in a deterministic order should keep everything deterministic. Maybe it even removes enough thread overhead in some scenarios to not be slower despite removing some parallelity.

I currently lack the C++ and CUDA proficiency, as well as the time to learn it to try that idea out myself; however, it may help someone in the future.
